maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java

```diff
@@ -2846,8 +2849,92 @@ public void addProviderProperties()
         @Nonnull
         public Set<Artifact> getProviderClasspath()
         {
+            String provider = "surefire-junit-platform";
             String version = surefireBooterArtifact.getBaseVersion();
-            return dependencyResolver.getProviderClasspath( "surefire-junit-platform", version );
+            Set<Artifact> providerArtifacts = dependencyResolver.getProviderClasspath( provider, version );
+            alignJUnitPlatformVersion( providerArtifacts );
+            resolveJUnitJupiterEngine( providerArtifacts );
+            resolveJUnitVintageEngine( providerArtifacts );
+            return providerArtifacts;
+        }
+
+        private void resolveJUnitJupiterEngine( Set<Artifact> providerArtifacts )
+        {
+            Artifact junitJupiterApi = getProjectArtifactMap().get( "org.junit.jupiter:junit-jupiter-api" );
+            if ( junitJupiterApi == null ) // no api, no engine
+            {
+                return;
+            }
+            Artifact junitJupiterEngine = getProjectArtifactMap().get( "org.junit.jupiter:junit-jupiter-engine" );
+            if ( junitJupiterEngine != null ) // engine already resolved by project
+            {
+                return;
+            }
+            // resolve "junit-jupiter-engine" and its transitive dependencies
+            String version = junitJupiterApi.getBaseVersion();
+            resolve( providerArtifacts, "org.junit.jupiter", "junit-jupiter-engine", version );
+        }
+
+        private void resolveJUnitVintageEngine( Set<Artifact> providerArtifacts )
+        {
+            Artifact junit = getProjectArtifactMap().get( "junit:junit" );
+            if ( junit == null || !junit.getBaseVersion().equals( "4.12" ) ) // no or wrong "junit", no engine
```

Comment:
The comments like this can be substituted with good name of `boolean` variable which expressed the purpose.
For instance, `boolean hasApplicableJUnit4InProject`.

Reply:
Even with the longer variable name I still see value in the comment. I even tend to split the `if`-clause into two distinct ones.
```java
if ( junit == null) { return; } // No JUnit at all? No vintage engine.
if ( !junit.getBaseVersion().equals( "4.12" ) ) { return; } // Not exactly JUnit 4.12? No vintage engine.
```

Reply:
I only wanted to say that it is a Clean Code to write expressive code for places where were previus comments.